Userland/Utilities/strace.cpp

```diff
@@ -36,24 +137,328 @@ static void handle_sigint(int)
     }
 }
 
-static String format_syscall(syscall_arg_t syscall_index, syscall_arg_t arg1, syscall_arg_t arg2, syscall_arg_t arg3, syscall_arg_t res)
+static void copy_from_process(const void* source_p, void* target_p, size_t size)
+{
+    const char* source = static_cast<const char*>(source_p);
+    char* target = static_cast<char*>(target_p);
+    for (size_t i = 0; i != ceil_div(size, sizeof(int)); i++) {
+        *((int*)target) = ptrace(PT_PEEK, g_pid, const_cast<void*>(static_cast<const void*>(source)), 0);
+        source += sizeof(int);
+        target += sizeof(int);
+    }
+}
+
+static ByteBuffer copy_from_process(const void* source, size_t length)
+{
+    auto buffer = ByteBuffer::create_uninitialized(length).value();
+    copy_from_process(source, buffer.data(), length);
+    return buffer;
+}
+
+template<typename T>
+static T copy_from_process(const T* source)
+{
+    T value {};
+    copy_from_process(source, &value, sizeof(T));
+    return value;
+}
+
+struct PointerArgument {
+    const void* value;
+};
+
+namespace AK {
+template<>
+struct Formatter<PointerArgument> : StandardFormatter {
+    Formatter() = default;
+    explicit Formatter(StandardFormatter formatter)
+        : StandardFormatter(formatter)
+    {
+    }
+
+    void format(FormatBuilder& format_builder, PointerArgument const& value)
+    {
+        auto& builder = format_builder.builder();
+        if (value.value == nullptr)
+            builder.append("null");
+        else
+            builder.appendff("{}", value.value);
+    }
+};
+}
+
+class FormattedSyscallBuilder {
+
+public:
+    FormattedSyscallBuilder(Syscall::Function syscall_function)
+    {
+        m_builder.append(Syscall::to_string(syscall_function));
+        m_builder.append('(');
+    }
+
+    template<typename T>
+    void add_argument(const char* format, T&& arg)
+    {
+        add_argument_separator();
+        m_builder.appendff(format, forward<T>(arg));
+    }
```

Comment:
Since you're using this as a general formatter:
```suggestion
    void add_argument(CheckedFormatString<T> format, T&& arg)
    {
        add_argument_separator();
        m_builder.appendff(format.view(), forward<T>(arg));
    }
```

To ensure that calls are checked at compiletime.

Reply:
Isn't that going to provoke two `CheckedFomatString<T>` creations? `StringBuilder.appendff` already takes a `CheckedFormatString<T>`, so there's already compile-time checking happening. What about this:

```diff
-    void add_argument(const char* format, T&& arg)
+    void add_argument(CheckedFormatString<RemoveCVReference<T>> format, T&& arg)
     {
         add_argument_separator();
-        m_builder.appendff(format, forward<T>(arg));
+        m_builder.appendff(move(format), forward<T>(arg));
```

Reply:
The constructors are `consteval`, so it makes no real difference.
you can move the object in, that's fine.

Also, please just pass `T` directly, no need for `RemoveCVReference`.

Reply:
I indeed tried with `T` directly first, but was getting problems because in some cases `T` was deduced to a reference type, sometimes also `const&`, hence the `RemoveCVReference`.

In any case I hadn't realised the `consteval`-ness of `CheckedFormatString` (again, I'm still making my way through the library...), so I'll go with the solution you provided.

Reply:
I indeed tried with `T` directly first, but was getting problems because in some cases `T` was deduced to a reference type, sometimes also `const&`, hence the `RemoveCVReference`.

In any case I hadn't realised the `consteval`-ness of `CheckedFormatString` (again, I'm still making my way through the library...), so I'll go with the solution you provided.

Reply:
Done.